File: 376-wiggle-subsequence/376-wiggle-subsequence.java
class Solution {
    public int wiggleMaxLength(int[] nums) {
        if (nums == null) return 0;
        if (nums.length <= 1) return nums.length;
        int f = 1, b = 1; //the first number can be a smaller number or larger number depends on what the next number it is. 
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > nums[i-1]) f = b + 1;
            else if (nums[i] < nums[i-1]) b = f + 1;
        }
        return Math.max(f, b);
    }
}

File: 376-wiggle-subsequence/376-wiggle-subsequence.cpp
class Solution {
public:
    int dp[1005];
    int solve(int i, int j, vector<int> nums)
    {
        if(i>=nums.size()) return 0;
        
        if(dp[j+1] != -1) return dp[j+1];
        
        int take = 0, nottake =  solve(i+1, j, nums);
        if(j==-1 || (nums[j]<0 && nums[i]>=0) || (nums[j]>=0 && nums[i]<=0))
            take = 1 + solve(i+1, i, nums);
        
        return dp[j+1] = max(take, nottake);
    }
    
    int wiggleMaxLength(vector<int>& nums) {
        if(nums.size()==1) return 1;

        vector<int> diff;
        
        for(int i=0; i<nums.size()-1; i++){
            if(nums[i+1]-nums[i] != 0)
                diff.push_back(nums[i+1]-nums[i]);
        }
        
        
        memset(dp, -1, sizeof(dp));
        return 1+solve(0,-1,diff);
        
        
    }
};